sympy/matrices/expressions/tests/test_matpow.py

```diff
@@ -39,3 +53,44 @@ def test_as_explicit_nonsquare():
     raises(ShapeError, lambda: MatPow(A, 2).as_explicit())
     raises(ShapeError, lambda: MatPow(A, -1).as_explicit())
     raises(ValueError, lambda: MatPow(A, pi).as_explicit())
+
+
+def test_doit_nonsquare_MatrixSymbol():
+    assert MatPow(A, 1).doit() == A
+    for r in [0, 2, -1, pi]:
+        assert MatPow(A, r).doit() == MatPow(A, r)
+
+
+def test_doit_square_MatrixSymbol_symsize():
+    assert MatPow(C, 0).doit() == Identity(n)
+    assert MatPow(C, 1).doit() == C
+    for r in [2, -1, pi]:
+        assert MatPow(C, r).doit() == MatPow(C, r)
+
+
+def test_doit_with_MatrixBase():
+    X = ImmutableMatrix([[1, 2], [3, 4]])
+    assert MatPow(X, 0).doit() == ImmutableMatrix(Identity(2))
+    assert MatPow(X, 1).doit() == X
+    assert MatPow(X, 2).doit() == X**2
+    assert MatPow(X, -1).doit() == X.inv()
+    assert MatPow(X, -2).doit() == (X.inv())**2
+    # less expensive than testing on a 2x2
+    assert MatPow(ImmutableMatrix([4]), S.Half).doit() == ImmutableMatrix([2])
+
+
+def test_doit_nonsquare():
+    # FIXME: or should these remain unevaluated?
+    X = ImmutableMatrix([[1, 2, 3], [4, 5, 6]])
+    assert MatPow(X, 1).doit() == X
+    raises(ShapeError, lambda: MatPow(X, 0).doit())
+    raises(ShapeError, lambda: MatPow(X, 2).doit())
+    raises(ShapeError, lambda: MatPow(X, -1).doit())
+    raises(ShapeError, lambda: MatPow(X, pi).doit())
```

Comment:
Oops, I left a FIXME in here.  I think these should evaluate and indeed should raise errors.  I will add a commit.

This is in contrast with `MatPow(<non-square-MatrixSymbol>, 0)` which should be unevaluated (and is tested above in `...nonsquare_MatrixSymbol`.  This is because the "n x m" MatrixSymbol to-zeroth-power could be ok (e.g., with a substitution) whereas the ImmutableMatrix is a particular matrix.  However, the logic is not perfect because `MatPow(MatrixSymbol(A, 2, 3), 0).doit()` does not raise an error.